### PR TITLE
Add overlay command to ease the creation of ext3 writable overlay

### PR DESCRIPTION
**Pull request by cclerget**
## Description of the Pull Request (PR):

Add overlay command to ease the creation of ext3 writable overlay
- allow to create an overlay as a single ext3 image file.
- allow to create and add an ext3 writable overlay to an existing SIF image.

SIF restrictions:
- SIF image must not contain a writable overlay
- SIF image must not be signed
- SIF image must not be encrypted

#### Before submitting a PR, make sure you have done the following:

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

**Comment by dtrudg** [comment]
Hi @cclerget - I'll probably get around to a proper review on Monday.

A quick thought though...

> SIF restrictions:
> 
>     * SIF image must not contain a writable overlay
> 
>     * SIF image must not be signed

Should there be some restriction or warning for SIFs where the rootfs is encrypted? Our overlay will not be, and that may not be intuitive / expected to the user?

**Comment by cclerget** [comment]
Ah right forgot encryption case, thanks !

File: cmd/internal/cli/overlay_create.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+)
+
+// -s|--size
+var overlaySizeFlag = cmdline.Flag{
+	ID:           "overlaySizeFlag",
+	Value:        &overlaySize,
+	DefaultValue: 64,
+	Name:         "size",
+	ShortHand:    "s",
+	Usage:        "size of the EXT3 writable overlay in MB",
```
Comment (dtrudg) [inline cmd/internal/cli/overlay_create.go]: Presumably this is MiB? If so can we put MiB here as we have had some confusion previously with powers of 10 MB vs powers of 2 MiB in some code.

File: docs/content.go [hunk 2]
```diff
@@ -1039,4 +1039,26 @@ Enterprise Performance Computing (EPC)`
 
   To display the resulting configuration instead of writing it to file:
   $ singularity config global --dry-run --set "bind path" /etc/resolv.conf`
+
+	OverlayUse   string = `overlay`
+	OverlayShort string = `Manage EXT3 writable overlay image`
```
Comment (dtrudg) [inline docs/content.go]: Either 'Manage **an** EXT3 image....' or ' Manage..... image**s**'

File: docs/content.go [hunk 3]
```diff
@@ -1039,4 +1039,26 @@ Enterprise Performance Computing (EPC)`
...
+	OverlayLong  string = `
+  The overlay command allows to manage EXT3 writable overlay image.`
```
Comment (dtrudg) [inline docs/content.go]: 'The overlay command allows management of EXT3 writable overlay images' ?

File: docs/content.go [hunk 4]
```diff
@@ -1039,4 +1039,26 @@ Enterprise Performance Computing (EPC)`
...
+  $ singularity overlay create --help`
+
+	OverlayCreateUse   string = `create <options> image`
+	OverlayCreateShort string = `Create EXT3 writable overlay image`
+	OverlayCreateLong  string = `
+  The overlay create command allows to create EXT3 writable overlay image either
+  as a single EXT3 image or by adding it automatically to an existing SIF image.`
+	OverlayCreateExample string = `
+  To create and add a writable overlay to an existing SIF image:
+  $ singularity overlay --size 1024 /tmp/image.sif
```
Comment (dtrudg) [inline docs/content.go]: Needs to be `overlay create` I think

File: docs/content.go [hunk 5]
```diff
@@ -1039,4 +1039,26 @@ Enterprise Performance Computing (EPC)`
...
+
+  To create a single EXT3 writable overlay image:
+  $ singularity overlay --size 1024 /tmp/my_overlay.img`
```
Comment (dtrudg) [inline docs/content.go]: Needs to be `overlay create` I think

File: internal/app/singularity/overlay_create.go [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+			sifArch = arch
+
+			img.File.Close()
+
+			for _, overlay := range overlays {
+				if overlay.Type != image.EXT3 {
+					continue
+				}
+				delCmd := fmt.Sprintf("singularity sif del %d %s", overlay.ID, imgPath)
+				return fmt.Errorf("an overlay writable partition already exist in %s (ID: %d), delete it first with %q", imgPath, overlay.ID, delCmd)
```
Comment (dtrudg) [inline internal/app/singularity/overlay_create.go]:
```suggestion
				return fmt.Errorf("a writable overlay partition already exists in %s (ID: %d), delete it first with %q", imgPath, overlay.ID, delCmd)
```

**Review by dtrudg — CHANGES_REQUESTED** [review]
Looks mostly good to me. A few minor docs / messages things commented here.

Couple of questions....

- Why not e.g. `truncate --size 1024M myfile` instead of `dd` as that will be faster and less likely to exhaust disk as it creates a sparse file?
- `mkfs.ext3` is not in the standard user PATH on at least debian 10+, as that doesn't include`/sbin`. Worth looking in `/sbin` by default?

**Comment by cclerget** [comment]
Thanks for the review @dtrudg , I applied your suggestions.

>     * Why not e.g. `truncate --size 1024M myfile` instead of `dd` as that will be faster and less likely to exhaust disk as it creates a sparse file?

IIRC sparse files have been source of bugs in v2, and it doesn't apply when adding overlay in a SIF image.
Also this is the way it is done in the documentation.

> 
>     * `mkfs.ext3` is not in the standard user PATH on at least debian 10+, as that doesn't include`/sbin`. Worth looking in `/sbin` by default?

Added `setPath` which is blinldy prepending the singularity default paths to the current PATH

**Review by dtrudg — APPROVED** [review]
LGTM

**Review by ikaneshiro — APPROVED** [review]

**Pull request merged**